petab/visualize/plot_data_and_simulation.py

```diff
@@ -125,18 +139,8 @@ def plot_data_and_simulation(
                                                    observable_id_list,
                                                    observable_num_list,
                                                    plotted_noise)
-
-    # import simulation file, if file was specified
-    if isinstance(sim_data, str):
-        sim_data = core.get_simulation_df(sim_data)
-
-    if DATASET_ID not in exp_data:
-        raise ValueError(f'Visualization requires field {DATASET_ID} to be  '
-                         f'present in measurement table.')
-
-    if sim_data is not None and DATASET_ID not in sim_data:
-        raise ValueError(f'Visualization requires field {DATASET_ID} to be '
-                         f'present in simulation table.')
+        if sim_data is not None:
+            sim_data[DATASET_ID] = exp_data[DATASET_ID]
```

Comment:
I feel like this line is just asking for trouble. It might work for that one specific use case but is generally not a good solution.
- you don't want to overwrite user provided value
- I don't think we can generally assume that sim_data and exp_data are equally ordered.

Also I generally recommend that if you fix something like this, add one or two test cases to make sure it is not inadvertently broken in the future.

Reply:
* `if sim_data is not None and DATASET_ID not in sim_data:` Hence this only applies if there is no user provided datasetId. In these cases exp_data gets a default datasetId (something like 'obsverbale_condition').
* I thought in cases without datasetId it would be reasonable to assume same order for sim_data and exp_data. But of course the procedure for creating exp_data's datasetId could be easily applied to sim_data, too.

Reply:
test wise this could be covered by enhancing each of the plot_by_observable/condition/datasetId tests with a second function call also plotting with simulated data and not only measurement

Reply:
> * `if sim_data is not None and DATASET_ID not in sim_data:` Hence this only applies if there is no user provided datasetId. In these cases exp_data gets a default datasetId (something like 'obsverbale_condition').

I can only see `if sim_data is not None` there. As a proposal for a fix this sounds right.

Reply:
ah yeah you are right. had to many intermediate versions.